nextcord/ui/modal.py

```diff
@@ -0,0 +1,406 @@
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING, Optional, List, Callable, Dict, Any, Tuple
+from functools import partial
+from itertools import groupby
+
+import sys
+import os
+import time
+import traceback
+import asyncio
+
+from ..utils import MISSING
+from .view import (
+    _ViewWeights,
+    _walk_all_components,
+    _component_to_item,
+)
+from .item import Item
+from ..components import Component, _component_factory
+
+__all__ = (
+    'Modal',
+    'ModalStore',
+)
+
+
+if TYPE_CHECKING:
+    from ..interactions import Interaction
+    from ..state import ConnectionState
+    from ..types.components import Component as ComponentPayload
+
+
+class Modal:
+    """Represents a discord modal popup.
+
+    This object must be inherited to create a modal within Discord.
+
+    .. versionadded:: 2.0
+
+    Parameters
+    -----------
+    title: :class:`str`
+        The title of the modal.
+    timeout: Optional[:class:`float`] = None
+        Timeout in seconds from last interaction with the UI before no longer accepting input.
+        If ``None`` then there is no timeout.
+    custom_id: :class:`str` = MISSING
+        The ID of the modal that gets received during an interaction.
+        If the ``custom_id`` is MISSING, then a random ``custom_id`` is set.
+    **components:
+        keyword-arguments of :class:`Item` representing the components of the
+        Modal where the kwarg value gets added as an attribute named to the
+        kwarg key.
+
+    Attributes
+    ------------
+    timeout: Optional[:class:`float`]
+        Timeout from last interaction with the UI before no longer accepting input.
+        If ``None`` then there is no timeout.
+    children: List[:class:`Item`]
+        The list of children attached to this view.
+    custom_id: str
+        The ID of the modal that gets received during an interaction.
+    """
+    
+    def __init__(
+        self,
+        title: str,
+        *,
+        timeout: Optional[float] = None,
+        custom_id: str = MISSING,
+        **components,
+    ):
+        self.title = title
+        self.timeout = timeout
+        self._provided_custom_id = custom_id is not MISSING
+        self.custom_id = os.urandom(16).hex() if custom_id is MISSING else custom_id
+        
+        self.children: List[Item] = []
+        for key, component in components.items():
+            setattr(self, key, component)
+            self.children.append(component)
+        
+        self.__weights = _ViewWeights(self.children)
```

Comment:
We can also create a ComponentsWeight class because the view and modals weight behave in exactly the same way

Reply:
View weight is required because each textinput is contained in an action row.
I don't think making two class for exactly the same thing is a good idea.

Reply:
Doesn't ViewWeights have hard coded limits like max 5 per row and max 25 total?

Reply:
Each component as a max length, and TextInput as a len of 5, like SelectMenu